gpinitsystem: Return 0 on warnings

slari-pivotal opened the pull request:
Previously gpinitsystem returned 1 in case of warngings (2 in case of
errors). We have changed gpinitsystem to return 0 for warnings so that other
tools using gpinitsystem (such as gpupgrade) will not fail when gpinitsystem
generates warnings.

Also, gpinitsystem was exiting with non-zero return code if certain errors or
warnings had happened on any previous run of gpinitsystem that day. This has
also been fixed in this commit

Co-authored-by: Kalen Krempely <kkrempely@pivotal.io>
Co-authored-by: Shoaib Lari <slari@pivotal.io>

## Here are some reminders before you submit the pull request

The corresponding 6X Backport is #8847

jchampio commented on gpMgmt/test/behave/mgmt_utils/gpinitsystem.feature:
```diff
@@ -17,6 +17,24 @@ Feature: gpinitsystem tests
+        Given create demo cluster config
+        # force a failure by passing no args
+        When the user runs "gpinitsystem"
+        Then gpinitsystem should return a return code of 2
```
Do we want to fix the `2` return code in this PR?

slari-pivotal replied:
Since `gpstop` also returns `2` on failure, may be we should have a cleanup story to make exit code consistent across our tools for GPDB 7 as there may be others also.

jchampio replied:
Maybe... the question in my mind is, what do we want the backported behavior of gpinitsystem in 6X to be when this code is enabled with the new flag? Do we really want to have "zero for okay and two for errors"?

slari-pivotal replied:
The `--ignore-warning` flag will make sure that we return 0 on warnings as well. We can increase the scope of the flag to return 1 instead of 2 on failures.  This would make the behavior even more incompatible with the present behavior. For 6X, it seems safer to limit the scope of the flag. However, if the clients/consumer utilities utilizing the new flag (i.e. gpupgrade) only want to check for 0 and 1 for success and failure, then we can return 1 instead of 2 for failures also.

jchampio replied:
> This would make the behavior even more incompatible with the present behavior. For 6X, it seems safer to limit the scope of the flag.

I don't think "degrees of incompatibility" are applicable here; it's either broken or it's not, and the break is hidden behind a flag. Any risk would come from the difficulty of implementation.

> However, if the clients/consumer utilities utilizing the new flag (i.e. gpupgrade) only want to check for 0 and 1 for success and failure, then we can return 1 instead of 2 for failures also.

Well, no, clients need to check for any non-zero error. I'm just confused as to why we'd select exit code `2` as the new error behavior. Once we backport the behavior it's locked for 6.

marksliva replied:
We'll change gpinitsystem to exit 1 on failure, and also backport this behavior when the `--ignore-warnings` flag is specified.

slari-pivotal replied:
We have changed gpinitsystem to return only 0 and 1.

slari-pivotal commented on gpMgmt/bin/gpinitsystem:
```diff
@@ -1475,8 +1468,7 @@ STOP_QD_PRODUCTION () {
 	    0 ) LOG_MSG "[INFO]:-Successfully shutdown the new Greenplum instance" ;;
-	    1 ) LOG_MSG "[WARN]:-Non fatal error from Greenplum instance shutdown, check log files, will continue"
-		EXIT_STATUS=1 ;;
+	    1 ) LOG_MSG "[WARN]:-Non fatal error from Greenplum instance shutdown, check log files, will continue" ;;
```
`gpstop` returns `2` on failure, and `1` when `gpstop -u` (to reload the pg_hba.conf parameters) fails for one or more segments.  So, that is why a return of `1` from gpstop is treated as a warning here.

jchampio replied:
`gpstop` also returns `1` if it doesn't run due to a bad `import` or Python syntax error -- unfortunately we don't have the power to declare `1` off-limits for other failure modes.

slari-pivotal replied:
We have changed gpinitsystem to return 1 if gpstop returns a non-zero return code.

slari-pivotal commented on gpMgmt/test/behave/mgmt_utils/environment.py:
```diff
@@ -18,7 +18,7 @@ def before_all(context):
 def before_feature(context, feature):
     # we should be able to run gpexpand without having a cluster initialized
-    tags_to_skip = ['gpexpand', 'gpaddmirrors', 'gpstate', 'gpmovemirrors', 'gpconfig', 'gpssh-exkeys', 'gpstop']
+    tags_to_skip = ['gpexpand', 'gpaddmirrors', 'gpstate', 'gpmovemirrors', 'gpconfig', 'gpssh-exkeys', 'gpstop', 'gpinitsystem']
```
Done.

kalensk commented:
In [gpcreateseg.sh](https://github.com/greenplum-db/gpdb/blob/8877a8e216748d1fabcb1aa33b9b1c2cd71c4ed0/gpMgmt/bin/lib/gpcreateseg.sh#L259) are we supposed to remove the `2` at the end of the line since we removed the second parameter from ERROR_EXIT?

`trap '$ECHO "KILLED:$SEGMENT_TO_CREATE" >> $PARALLEL_STATUS_FILE;ERROR_EXIT "[FATAL]:-[$INST_COUNT]-Recieved INT or TERM signal"' INT TERM`?

kalensk commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -626,9 +618,9 @@ GET_MASTER_PORT () {
+			ERROR_EXIT "[FATAL]:-MASTER_DATA_DIRECTORY variable not set" 1;fi
 		if [ ! -d $MASTER_DATA_DIRECTORY ]; then
-				ERROR_EXIT "[FATAL]:-No $MASTER_DATA_DIRECTORY directory" 2
+				ERROR_EXIT "[FATAL]:-No $MASTER_DATA_DIRECTORY directory" 1
```
Seems like we are passing `1` as a second parameter to ERROR_EXIT here, but I thought we removed the second parameter from ERROR_EXIT?

slari-pivotal replied:
Fixed.

kalensk commented on gpMgmt/test/behave/mgmt_utils/gpinitsystem.feature:
```diff
@@ -25,8 +78,8 @@ Feature: gpinitsystem tests
         And the user runs command "rm -rf /tmp/gpinitsystemtest && mkdir /tmp/gpinitsystemtest"
         # stop db and make sure cluster config exists so that we can manually initialize standby
         And the cluster config is generated with data_checksums "1"
-        When the user runs "gpinitsystem -a -c ../gpAux/gpdemo/clusterConfigFile -l /tmp/gpinitsystemtest -s localhost -P 21100 -S /wrong/path -h ../gpAux/gpdemo/hostfile"
```
Probably super obvious, but why was `-l /tmp/gpinitsystemtest` removed? I did notice the new step `create demo cluster config` which is nice.

slari-pivotal replied:
The `SCAN_LOG()` function scans the log directory for any warnings. Instead of logging to a temporary directory, we go back to default. This way, if this test had a side-effect on the next test, we will catch it.

dkrieger-pvtl commented:
https://github.com/marksliva/gpdb/blob/gpinitsystem-exit-code-168785723/gpMgmt/bin/lib/gp_bash_functions.sh#L1076 should have a space removed.
```
	        \/
        else
		 LOG_MSG "[WARN]:-Could not locate status file $1" 1
		REPORT_FAIL=1
	fi
```

slari-pivotal commented:
> https://github.com/marksliva/gpdb/blob/gpinitsystem-exit-code-168785723/gpMgmt/bin/lib/gp_bash_functions.sh#L1076 should have a space removed.
> 
> ```
> 	        \/
>         else
> 		 LOG_MSG "[WARN]:-Could not locate status file $1" 1
> 		REPORT_FAIL=1
> 	fi
> ```

Fixed.

slari-pivotal commented:
> In [gpcreateseg.sh](https://github.com/greenplum-db/gpdb/blob/8877a8e216748d1fabcb1aa33b9b1c2cd71c4ed0/gpMgmt/bin/lib/gpcreateseg.sh#L259) are we supposed to remove the `2` at the end of the line since we removed the second parameter from ERROR_EXIT?
> 
> `trap '$ECHO "KILLED:$SEGMENT_TO_CREATE" >> $PARALLEL_STATUS_FILE;ERROR_EXIT "[FATAL]:-[$INST_COUNT]-Recieved INT or TERM signal"' INT TERM`?

I have removed the extra parameter.

jchampio commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -568,11 +562,11 @@ GET_REPLY () {
 	if [ $REPLY != Y ] && [ $REPLY != y ]; then
 		LOG_MSG "[WARN]:-User abort requested, Script Exits!" 1
-		exit 1
+		exit 0
```
I know the log messages say `WARN`, but do we want to claim successful execution if the user aborts?

slari-pivotal replied:
I think we should be consistent in returning 0 for `WARN` and 1 for `FATAL`. If we want to return 1 for the abort case, then we should replace `WARN` by `FATAL`.

kalensk replied:
I agree with @jchampio in that we should be returning exit 1 and for consistency update the  LOG_MSG to `FATAL`. A fixup has been made.
